src/parse/mod.rs

```diff
@@ -79,38 +79,23 @@ where
     ///     If there is at least one token that is not an error, returns a parser.
     ///     Otherwise, returns a list of the errors.
     /// Otherwise, returns None.
-    pub fn new(mut iter: I, debug: bool) -> Result<Self, VecDeque<CompileError>> {
-        let mut error_handler = ErrorHandler::new();
-        let mut pending = VecDeque::new();
-        let first = loop {
-            match iter.next() {
-                Some(Ok(token)) => break token,
-                Some(Err(err)) => pending.push_back(err),
-                None if pending.is_empty() => {
-                    error_handler.push_back(CompileError::semantic(Locatable::new(
-                        "cannot have empty program".to_string(),
-                        Default::default(),
-                    )));
-                    return Err(pending);
-                }
-                None => return Err(pending),
-            }
-        };
-        if !pending.is_empty() {
-            return Err(pending);
-        }
-        Ok(Parser {
-            scope: Scope::new(),
-            tag_scope: Scope::new(),
-            tokens: iter,
+    pub fn new(first: Locatable<Token>, tokens: I, debug: bool) -> Self {
```

Comment:
Why do you have to pass the first token to `Parser::new`? Why can't it just call next on tokens itself?

Reply:
I now realize that this is so that you can catch early lexer errors in the compile function. See comment in src/lib.rs

Reply:
This is the whole point of the change. If it calls `next` on tokens itself, then `new()` has to return a `Result<_, Vec<CompileError>>`, there's no way to tell it 'no I'm sure this first token is valid'. See [`cpp`](https://github.com/jyn514/rcc/commit/80ee39f05efbb928852a5535bb7b8c2432ab2779#diff-a165b99f3b4afa48136cbfc5113ee601R171) for why this might be useful. Note also there that there's no way to return a `Vec` of errors, I'd have to add a new `error_handler` (although I'll probably have to end up doing that anyway).

Now that I think about it, maybe this isn't worth it ...